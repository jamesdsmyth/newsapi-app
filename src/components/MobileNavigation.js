import React from 'react';
import PropTypes from 'prop-types';

const MobileNavigation = ({ showNavigation, hideNavigation }) => (
  <section className="header-content__mobile-buttons">
    <span
      className="header-content__mobile-buttons__menu"
      onClick={showNavigation}>
      Menu
    </span>
    <span
      className="header-content__mobile-buttons__menu__hide"
      onClick={hideNavigation}>
      x
    </span>
  </section>
);

MobileNavigation.propTypes = {
  showNavigation: PropTypes.func.isRequired,
  hideNavigation: PropTypes.func.isRequired
}

export default MobileNavigation;