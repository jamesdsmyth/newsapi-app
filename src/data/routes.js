// routes - the categories available from newsapi.org
const routes = [
  {
    route: 'home',
    category: 'general'
  },
  {
    route: 'business',
    category: 'business'
  },
  {
    route: 'entertainment',
    category: 'entertainment'
  },
  {
    route: 'health',
    category: 'health'
  },
  {
    route: 'science',
    category: 'science'
  },
  {
    route: 'sports',
    category: 'sports'
  },
  {
    route: 'technology',
    category: 'technology'
  }
]

export default routes;