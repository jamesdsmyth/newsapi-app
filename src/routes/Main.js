import React, { Component } from 'react';

import { Route, Redirect } from "react-router-dom";

import NewsLanding from '../containers/NewsLanding';
import NewsArticleLanding from '../containers/NewsArticleLanding';

// Main is the container that houses the routes of the app
class Main extends Component {
  render() {
    return (
      <main className="main">
        <Route exact path="/" render={() => (
            <Redirect to="/home"/>
          )}
        />
        {
          this.props.routes.map((item, index) => (
            <Route
              exact
              path={`/${item.route}`}
              key={index}
              render={() => (
                <NewsLanding category={item.category} />
              )}
            />
          ))
        }
        {
          this.props.routes.map((item, index) => (
            <Route
              path={`/${item.route}/:articleId`}
              key={index}
              render={(props) => (
                <NewsArticleLanding {...props} category={item.category} />
              )}
            />
          ))
        }
      </main>
    )
  }
}

export default Main;