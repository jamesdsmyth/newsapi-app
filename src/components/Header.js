import React, { Component } from 'react';
import Navigation from './Navigation';
import MobileNavigation from './MobileNavigation';

class Header extends Component {

  constructor() {
    super();

    this.state = {
      headerHidden: true
    }

    this.hideNavigation = this.hideNavigation.bind(this);
    this.showNavigation = this.showNavigation.bind(this);
  }

  // hiding the mobile navigation
  hideNavigation() {
    this.setState({
      headerHidden: true
    });

    window.scrollTo(0, 0);
  }

  // showing the mobile navigation
  showNavigation() {
    this.setState({
      headerHidden: false
    });
  }

  render() {
    return (
      <header className={`header-content ${this.state.headerHidden ? `header-content--hidden` : ``}`}>
        <MobileNavigation 
          hideNavigation={this.hideNavigation}
          showNavigation={this.showNavigation}
        />
        <Navigation 
          routes={this.props.routes}
          hideNavigation={this.hideNavigation}
        />
      </header>
    )
  }
}

export default Header;