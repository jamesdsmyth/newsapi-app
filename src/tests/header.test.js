import React from 'react';
import { mount, shallow, configure } from 'enzyme';
import expect from 'expect';
import sinon from 'sinon';
import Header from '../components/Header';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';

import MobileNavigation from '../components/MobileNavigation';

import routes from '../data/routes';

configure({ adapter: new Adapter() });

describe('<Header />', () => {

  const wrapper = shallow(
    <Router>
      <Header routes={routes} />
    </Router>
  );

  const component = renderer.create(
    <Router>
      <Header routes={routes} />
    </Router>
  );

  const headerComponent = shallow(
    <Header routes={routes} />
  );

  it('should render the header section', () => {
    expect(wrapper.find('.header-content')).toBeDefined();
  });

  it('should render the mobile button section', () => {
    expect(wrapper.find('.header-content__mobile-buttons')).toBeDefined();
  });

  it('should render the mobile button menu button', () => {
    expect(wrapper.find('.header-content__mobile-buttons__menu')).toBeDefined();
  });

  it('renders <Header /> and matches the snapshot', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  test('render the initial state of headerHidden as true', () => {
    expect(headerComponent.state().headerHidden).toBe(true);
  });

  test('the click event on the show menu button and toggle the state', () => {
    const callback = sinon.spy();
    const componentInstance = headerComponent.instance();
    const mobileNavigation = shallow(
      <MobileNavigation 
        hideNavigation={callback}
        showNavigation={callback}
      />
    );
    
    expect(headerComponent.find('.header-content').hasClass('header-content--hidden')).toBe(true);
    
    mobileNavigation.find('.header-content__mobile-buttons__menu').simulate('click');
    expect(callback.calledOnce).toBe(true);

    componentInstance.showNavigation();

    expect(headerComponent.state().headerHidden).toBe(false);
    expect(headerComponent.find('.header-content').hasClass('header-content--hidden')).toBe(false);
  });

  test('the click event on the hide menu button and toggle the state', () => {
    const callback = sinon.spy();
    const componentInstance = headerComponent.instance();
    window.scrollTo = jest.fn();
    const mobileNavigation = mount(
      <MobileNavigation 
        hideNavigation={callback}
        showNavigation={callback}
      />
    );

    expect(headerComponent.find('.header-content').hasClass('header-content--hidden')).toBe(false);

    mobileNavigation.find('.header-content__mobile-buttons__menu__hide').simulate('click');

    expect(callback.calledOnce).toBe(true);

    componentInstance.hideNavigation();

    expect(headerComponent.state().headerHidden).toBe(true);
    expect(headerComponent.find('.header-content').hasClass('header-content--hidden')).toBe(true);

    window.scrollTo.mockClear();
  });
});