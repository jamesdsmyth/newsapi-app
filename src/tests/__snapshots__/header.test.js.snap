// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders <Header /> and matches the snapshot 1`] = `
<header
  className="header-content header-content--hidden"
>
  <section
    className="header-content__mobile-buttons"
  >
    <span
      className="header-content__mobile-buttons__menu"
      onClick={[Function]}
    >
      Menu
    </span>
    <span
      className="header-content__mobile-buttons__menu__hide"
      onClick={[Function]}
    >
      x
    </span>
  </section>
  <nav
    className="navigation"
  >
    <ul
      className="navigation__main"
    >
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/home"
          onClick={[Function]}
        >
          home
        </a>
      </li>
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/business"
          onClick={[Function]}
        >
          business
        </a>
      </li>
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/entertainment"
          onClick={[Function]}
        >
          entertainment
        </a>
      </li>
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/health"
          onClick={[Function]}
        >
          health
        </a>
      </li>
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/science"
          onClick={[Function]}
        >
          science
        </a>
      </li>
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/sports"
          onClick={[Function]}
        >
          sports
        </a>
      </li>
      <li
        className="navigation__main__list-item"
        onClick={[Function]}
      >
        <a
          aria-current={null}
          className="navigation__main__list-item__link"
          href="/technology"
          onClick={[Function]}
        >
          technology
        </a>
      </li>
    </ul>
  </nav>
</header>
`;
